be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -1,86 +1,170 @@
-      receivedResponse.keys.foreach(channel => {
-        receivedResponse(channel).foreach(message => {
-          writeOnDb(channel, message, dbActorRef, MAX_ATTEMPT)
-        })
-      })
-      Right(jsonRpcMessage)
+      val validatorFlow = validator(mediatorActorRef, messageRegistry)
+      val receivedResponse: Map[Channel, Set[GraphMessage]] =
+        jsonRpcMessage.result.get
+          .resultMap.get
+          .map {
+            case (channel, set) => (channel, set.map(msg => Right(JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.PUBLISH, new Publish(channel, msg), Some(0)))))
+          }
+
+      val failedMessages = passThroughPipeline(receivedResponse, validatorFlow, MAX_RETRY)
+      if (failedMessages.values.forall(set => set.isEmpty)) {
+        Right(JsonRpcResponse(RpcValidator.JSON_RPC_VERSION, new ResultObject(0), None))
+      } else {
+        Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "GetMessagesById handler failed to process some messages", None))
+      }
+
+    // Will end up in a Sink.ignore, safe to let through
     case value @ _ => value
-  }.filter(_ => false) // instead of implementing a Sink[GraphMessage], we chose to implement it as a Flow and filter every single outgoing graphMessage
+  }
 
+  // This function will try to digest the set of messages for an entire channel
+  // It will go channel by channel until it finished processing messages or MAX_RETRY is reached
   @tailrec
-  private def writeOnDb(channel: Channel, message: Message, dbActorRef: AskableActorRef, remainingAttempts: Int): Unit = {
-    if (remainingAttempts != 0) {
-      val ask = dbActorRef ? DbActor.WriteAndPropagate(channel, message)
-      Await.ready(ask, duration).value match {
-        case Some(Failure(_: DbActorNAckException)) =>
-          writeOnDb(channel, message, dbActorRef, remainingAttempts - 1)
-        case _ =>
-      }
+  private def passThroughPipeline(receivedResponse: Map[Channel, Set[GraphMessage]], validatorFlow: Flow[GraphMessage, GraphMessage, NotUsed], remainingAttempts: Int)(implicit system: ActorSystem): Map[Channel, Set[GraphMessage]] = {
+    if (receivedResponse.isEmpty || remainingAttempts <= 0)
+      return receivedResponse
+
+    var failedMessages: Map[Channel, Set[GraphMessage]] = Map.empty
+    receivedResponse.foreach {
+      case (channel, messagesSet) =>
+        val result = messageThroughPipeline(validatorFlow, messagesSet)
+        failedMessages += channel -> result._1
+        // only log a during last attempt
+        if (remainingAttempts == 1 && result._2.nonEmpty) {
+          println("Errors: " + result._2)
+        }
     }
+
+    passThroughPipeline(failedMessages, validatorFlow, remainingAttempts - 1)
   }
 
+  // Push a single message through the pipeline
+  private def messageThroughPipeline(validator: Flow[GraphMessage, GraphMessage, NotUsed], messageSet: Set[GraphMessage])(implicit system: ActorSystem): (Set[GraphMessage], List[PipelineError]) = {
```

Comment:
It pushes all the messages in `messageSet` so it could be renamed `messagesThroughPipeline` and the documentation should be updated